File: dynamic-programming/partitionEqualSubsetSum.js
// LeetCode #416: https://leetcode.com/problems/partition-equal-subset-sum/

// O(n^2)  
let dp = function(nums) {
  let sum = 0
  for (let num of nums) {
    sum += num
  }
  
  if (sum % 2 != 0) {
    return false
  }
  
  sum = sum / 2

  let firstSet = new Set()
  firstSet.add(0)
  for (let i = 0; i < nums.length; i++) {
    let secondSet = new Set(firstSet)
    for (let val of firstSet.values()) {
      secondSet.add(val + nums[i])
    }

    if (secondSet.has(sum)) {
      return true;
    }
    
    firstSet = secondSet
  }
  
  return false;
};

// O(2^n), trying every combination, basically just power set generation
let bruteForce = function(nums) {
  let sum = 0
  for (let num of nums) {
    sum += num
  }
  
  if (sum % 2 != 0) {
    return false
  }
  
  sum = sum / 2

  let counter = Math.pow(2, nums.length)
  //check all combinations of array
  for (let i = 0; i < counter; i++) {
    let tmpSum = 0
    for (let j = 0; j < nums.length; j++) {
      if (i & (1 << j)) {
        tmpSum += nums[j]
      }
    }
    
    if (tmpSum == sum) {
      return true;
    }
  }
  
  return false;
};
}

export { bruteForce, dp }

File: dynamic-programming/partition_equal_subset_sum.rs
use std::collections::HashSet;

impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let mut sum = 0;
        for num in nums.iter() {
            sum += num;
        }
        
        if sum % 2 != 0 {
            return false;
        }
        
        sum /= 2;
        
        let mut first_set = HashSet::new();
        first_set.insert(0);
        for i in 0..nums.len() {
            let mut second_set = HashSet::new();
            for val in first_set.iter() {
                second_set.insert(val + nums[i]);
            }
            
            if second_set.contains(&sum) {
                return true;
            }
            
            for val in second_set.iter() {
                first_set.insert(*val);
            }
        }
        
        false
    }
  
    pub fn brute_force(nums: Vec<i32>) -> bool {
        false
    }
}
